sdks/java/io/kudu/src/main/java/org/apache/beam/sdk/io/kudu/KuduIO.java

```diff
@@ -0,0 +1,477 @@
+ *
+ * <h3>Reading from Kudu</h3>
+ *
+ * <p>{@code KuduIO} provides a source to read and returns a bounded collection of entities as
+ * {@code PCollection&lt;T&gt;}. An entity is built by parsing a Kudu {@link RowResult} using the
+ * provided {@link SerializableFunction&lt;RowResult, T&gt;}.
+ *
+ * <p>The following example illustrates various options for configuring the IO:
+ *
+ * <pre>{@code
+ * pipeline.apply(
+ *     KuduIO.<String>read()
+ *         .withMasterAddresses("kudu1:8051,kudu2:8051,kudu3:8051")
+ *         .withTable("table")
+ *         .withParseFn(
+ *             (SerializableFunction<RowResult, String>) input -> input.getString(COL_NAME))
+ *         .withCoder(StringUtf8Coder.of()));
+ *     // above options illustrate a typical minimum set, returns PCollection<String>
+ * }</pre>
+ *
+ * <p>{@code withCoder(...)} may be omitted if it can be inferred from the @{CoderRegistry}.
+ * However, when using a Lambda Expression or an anonymous inner class to define the function, type
+ * erasure will prohibit this. In such cases you are required to explicitly set the coder as in the
+ * above example.
+ *
+ * <p>Optionally, you can provide {@code withPredicates(...)} to apply a query to filter rows from
+ * the kudu table.
+ *
+ * <p>Optionally, you can provide {@code withProjectedColumns(...)} to limit the columns returned
+ * from the Kudu scan to improve performance. The columns required in the {@code ParseFn} must be
+ * declared in the projected columns.
+ *
+ * <p>Optionally, you can provide {@code withBatchSize(...)} to set the number of bytes returned
+ * from the Kudu scanner in each batch.
+ *
+ * <p>Optionally, you can provide {@code withFaultTolerent(...)} to enforce the read scan to resume
+ * a scan on another tablet server if the current server fails.
+ *
+ * <h3>Writing to Kudu</h3>
+ *
+ * <p>The Kudu sink executes a set of operations on a single table. It takes as input a {@link
+ * PCollection PCollection&lt;T&gt;} and a {@link FormatFunction&lt;T&gt;} which is responsible for
+ * converting the input into an idempotent transformation on a row.
+ *
+ * <p>To configure a Kudu sink, you must supply the Kudu master addresses, the table name and a
+ * {@link FormatFunction} to convert the input records, for example:
+ *
+ * <pre>{@code
+ * PCollection<MyType> data = ...;
+ * FormatFunction<MyType> fn = ...;
+ *
+ * data.apply("write",
+ *     KuduIO.write()
+ *         .withMasterAddresses("kudu1:8051,kudu2:8051,kudu3:8051")
+ *         .withTable("table")
+ *         .withFormatFn(fn));
+ * }</pre>
+ *
+ * <h3>Experimental</h3>
+ *
+ * {@code KuduIO} does not support authentication in this release.
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class KuduIO {
+  private static final Logger LOG = LoggerFactory.getLogger(KuduIO.class);
+
+  private KuduIO() {}
+
+  public static <T> Read<T> read() {
+    return new AutoValue_KuduIO_Read.Builder<T>().setKuduService(new KuduServiceImpl<>()).build();
+  }
```

Comment:
I would recommend have some common helper functions here so that Coders don't need to be always set (e.g. readBytes -> byte[], readStrings -> String, etc.). However this can be done in a later PR